fix ManagedExecutorConfig and ThreadContextConfig CDI qualifiers to be Dependent scoped and state when the same instance is returned

njr-11 opened the pull request:
pull fixes #40

Signed-off-by: Nathan Rauh <nathan.rauh@us.ibm.com>

njr-11 commented on api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorConfig.java:
```diff
@@ -19,27 +19,42 @@
+@Qualifier
 @Retention(RUNTIME)
-@Target(FIELD)
+@Target({ FIELD, METHOD, PARAMETER, TYPE })
```
I couldn't think of a reason for wanting TYPE either, but added it after reviewing documentation such as 
https://javaee.github.io/tutorial/cdi-basic006.html
and
https://docs.oracle.com/javaee/6/tutorial/doc/gjbck.html
which states that, "A qualifier type is a Java annotation defined as @Target({METHOD, FIELD, PARAMETER, TYPE}) and @Retention(RUNTIME)."
This seemed to suggest TYPE is required for qualifiers.
However, it appears [other documentation](https://docs.jboss.org/cdi/spec/1.2/cdi-spec.html) uses the term "typically", meaning it isn't a requirement: 
"A qualifier type is a Java annotation defined as @Retention(RUNTIME). Typically a qualifier type is defined as @Target({METHOD, FIELD, PARAMETER, TYPE})."

Let's have Matej, who is an expert in CDI, and also involved in this spec review this.

manovotn replied:
You are _not required_ to have `TYPE` there.
CDI specification uses the word ["typically"](http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#defining_qualifier_types) because most of the time you want to be able to place such qualifier on a bean backed by a Java class.

It is completely up to you what do you actually restrict it to. Then again, there is no real drawback to keeping `TYPE` there and you may, theoretically, have a class using it (a implementation of `ManagedExecutor` with this qualifier on it).

njr-11 commented:
@manovotn Could you review this pull for fixing our ManagedExecutorConfig/ThreadContextConfig qualifier annotations, specifically the concern raised by Andy around the proper list of targets? We are seeing some inconsistent language between different documents and aren't certain whether TYPE is required/needed. Thanks!

njr-11 commented:
@manovotn I replaced the commit under this pull with one that follows the approach you outlined.  Could you review again?  Thanks!

Also, I just noticed that you left a 'suggested changes' that I completely overlooked, and is no longer accessible probably because I removed my original commit.  Sorry about that.  Hopefully the commit that I replaced it with is consistent and covers everything correctly.

manovotn commented on api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorConfig.java:
```diff
@@ -19,27 +19,39 @@
  * </code></pre>
  *
- * <p>A <code>ManagedExecutor</code> must fail to inject, raising
+ * <p>A {@link ManagedExecutor}'s life cycle is tied to that of the application.
```
Since it is a CDI bean with dependent scope, the lifecycle of such bean is tied to the bean into which it is injected. 
E.g. if you inject it into you `Foo` bean and the context in which `Foo` sits is destroyed, then CDI attempts to destroy this dependent bean as well (calling PreDestroy/Dispose methods). Now this is effectively overriden by the enforced sharing as you keep a reference to that instance hence it stays alive. This isn't a problem but the sentence is somewhat misleading.

How about something like this: `All created instances of ManagedExecutor are destroyed when the application is stopped and the container then automatically shuts down all managed executors and cancels its remaining actions/tasks`

njr-11 replied:
I like your proposed changing in wording. I'll switch it to that.

manovotn commented on api/src/main/java/org/eclipse/microprofile/concurrent/ThreadContextConfig.java:
```diff
@@ -19,27 +19,42 @@
  * ...
  * </code></pre>
  *
+ * <p>A {@link ThreadContext} service's life cycle is tied to that of the application.
```
Similar comment as above. Also here is a mention of a `service` while there is no such mention with ManagerExecutor - we should either speak about services in both cases or in neither. Personally I wouldn't mention it as a service.

njr-11 replied:
Agree.  I think 'service' came from thinking of EE Concurrency's ContextService, which parallel's MicroProfile Concurrency's ThreadContext, but I'll remove the term because this is the MicroProfile spec, not the EE one.

njr-11 commented on api/src/main/java/org/eclipse/microprofile/concurrent/ManagedExecutorConfig.java:
```diff
@@ -19,27 +19,39 @@
+ * <p>Provides configuration for an injected {@link ManagedExecutor} instance
+ * which is {@link javax.enterprise.context.Dependent dependent} scoped.
+ * When an application has multiple injection points for {@link ManagedExecutor}
+ * with matching configuration, the container must inject the same instance.
```
If the application is to rely on the concurrency constraints provided by the managed executor configuration, it needs a guarantee that those constraints will be enforced and will not vary by implementation.  This targets the most common pattern, but I do agree that a valid pattern also exists where an application might want duplicate executors with identical configuration.  For this less common case, they would currently only be able to create one of those via CDI and would need to rely on the builder to create the other (with a producer, they could make it available in CDI as well).  Another way to address this could be to add a 'name' attribute to the configuration that would allow for arbitrary distinction between instances. 
`@Inject @ManagedExecutorConfig(name="high-priority-work", maxAsync=5)`
`@Inject @ManagedExecutorConfig(name="low-priority-work", maxAsync=5)`
 Do you have any insights on how this is typically addressed in CDI?

manovotn replied:
The name would work, inside the producer methods providing this bean, you could determine whether you create a new one or return an existing instance. Not sure if it is worth the hassle though, we probably don't have use case for this yet.
So we could just keep a GH issue on the side track and drop the idea for now.

njr-11 commented:
Regarding the general note about life cycle, I agree it is a concern (even more so for the builders) and have been thinking about how to address it.  The disposer method approach is a good idea but unfortunately has some drawbacks.  It doesn't cover instances created by the builders, which are the bigger problem.  Also, the fact that a bean instance has gone away doesn't mean that the CompletableFuture it created is no longer needed elsewhere in the application.  One of the driving reasons behind the addition of the spec to MicroProfile is the ability to return CompletableFuture from Fault Tolerance ```@Asynchronous```.  Even if we also add logic that detects that all of the CompletableFuture instances and their whole chain of dependent stages are complete, we still don't know whether the application will subsequently chain additional dependent stages.  This requires the continued availability of the managed executor.
Some other ideas:
It might be possible to tie into garbage collection to shut down managed executors when there are no more references held from CompletableFutures or other application code.
Otherwise, I think we need a manual way for the user to indicate they are done with a managed executor.  The most obvious way would be to lift the spec-imposed restriction against invoking the shutdown() API.  However, I'm hesitant to do that because we need to preserve consistency with EE Concurrency, which explicitly forbids the executor shutdown methods.  It might be possible to make the argument that we will allow shutdown() for managed executors created specifically via the new mechanisms in MicroProfile Concurrency and that therefore it is still consistent with EE Concurrency.
That said, this particular topic (shut down) ought to be a separate issue, and I'll get one opened for it.

njr-11 commented:
@manovotn thanks for the review comments.  I've made the requested fixes under [this commit](https://github.com/eclipse/microprofile-concurrency/pull/42/commits/624ae132b985e6e6a3e3c0dc5789df200167e585).  Please review.
Also note that I'm opening a separate issue for shutdown of managed executors because it's a bigger problem that even more so applies to instances created with the builders than CDI.

manovotn commented:
Good point with the builders and thanks for that new issue.
As for this PR, I think I ran out of things to nag you with :)

njr-11 commented on api/src/main/java/org/eclipse/microprofile/concurrent/ThreadContextConfig.java:
```diff
@@ -38,8 +45,9 @@
+@Qualifier
 @Retention(RUNTIME)
-@Target(FIELD)
+@Target({ FIELD, METHOD, PARAMETER, TYPE })
```
resolved in prior comment discussion
